cupy/cuda/cufft.pyx

```diff
@@ -105,17 +154,133 @@ cpdef size_t getVersion() except? -1:
+
+    # allocate another buffer to prepare for order conversion
+    for i in range(nGPUs):
+        gpus.append(arr.descriptor.GPUs[i])
+        sizes.append(arr.descriptor.size[i])
+    temp_xtArr, temp_xtArr_buffer = _XtMalloc(gpus, sizes,
+                                              CUFFT_XT_FORMAT_INPLACE)
+    temp_arr = <XtArray*>temp_xtArr
+
+    # Make a device copy to bring the data from the permuted order back to
+    # the natural order. Note that this works because after FFT
+    # arr.subFormat is silently changed to CUFFT_XT_FORMAT_INPLACE_SHUFFLED
+    with nogil:
+        result = cufftXtMemcpy(plan, <void*>temp_arr, <void*>arr,
+                               CUFFT_COPY_DEVICE_TO_DEVICE)
+    check_result(result)
+
+    for i in range(nGPUs):
+        # swap MemoryPointer in xtArr_buffer
+        temp = temp_xtArr_buffer[i]
+        temp_xtArr_buffer[i] = xtArr_buffer[i]
+        xtArr_buffer[i] = temp
+
+        # swap pointer in xtArr
+        arr.descriptor.data[i] = temp_arr.descriptor.data[i]
+        assert arr.descriptor.size[i] == temp_arr.descriptor.size[i]
+        temp_arr.descriptor.data[i] = NULL
+        temp_arr.descriptor.size[i] = 0
+
+    # temp_xtArr now points to the old data, which is now in temp_xtArr_buffer
+    # and will be deallocated after this line (out of scope)
+    _XtFree(temp_xtArr)
+
+
+# We need to manage the buffers ourselves in order to 1. avoid excessive,
+# uncessary memory usage, and 2. use CuPy's memory pool.
+# This is meant to replace cufftXtMalloc().
+cdef _XtMalloc(list gpus, list sizes, XtSubFormat fmt):
+    cdef XtArrayDesc* xtArr_desc
+    cdef XtArray* xtArr
+    cdef list xtArr_buffer = []
+    cdef int i, nGPUs
+    cdef size_t size
+
+    assert len(gpus) == len(sizes)
+    nGPUs = len(gpus)
+    xtArr_desc = <XtArrayDesc*>PyMem_Malloc(sizeof(XtArrayDesc))
+    xtArr = <XtArray*>PyMem_Malloc(sizeof(XtArray))
+    c_memset(xtArr_desc, 0, sizeof(XtArrayDesc))
+    c_memset(xtArr, 0, sizeof(XtArray))
+
+    xtArr_desc.nGPUs = nGPUs
+    for i, (gpu, size) in enumerate(zip(gpus, sizes)):
+        with Device(gpu):
+            buf = memory.alloc(size)
+        assert gpu == buf.device_id
+        xtArr_buffer.append(buf)
+        xtArr_desc.GPUs[i] = gpu
+        xtArr_desc.data[i] = <void*>buf.ptr
+        xtArr_desc.size[i] = size
+
+    xtArr.descriptor = xtArr_desc
+    xtArr.subFormat = fmt
+
+    return <intptr_t>xtArr, xtArr_buffer
+
+
+# This is meant to replace cufftXtFree().
+# We only free the C structs. The underlying GPU buffers are deallocated when
+# going out of scope.
+cdef _XtFree(intptr_t ptr):
+    cdef XtArray* xtArr = <XtArray*>ptr
+    cdef XtArrayDesc* xtArr_desc = xtArr.descriptor
+    PyMem_Free(xtArr_desc)
+    PyMem_Free(xtArr)
+
+
 class Plan1d(object):
-    def __init__(self, int nx, int fft_type, int batch):
+    def __init__(self, int nx, int fft_type, int batch, *,
+                 use_multi_gpus=False, devices=None, out=None):
```

Comment:
`use_multi_gpus` keyword argument seems redundant.

Reply:
I feel a little confused that the keyword argument `devices` is interpreted as the number of devices if an integer is given, and as the device ids if a tuple is given.
How about taking a single device id or a tuple of device ids?

Reply:
> `use_multi_gpus` keyword argument seems redundant.

No, it's used to split the code path within `Plan1d`. The required steps to initialize a single-GPU plan is different from those for a multi-GPU one.

> I feel a little confused that the keyword argument `devices` is interpreted as the number of devices if an integer is given, and as the device ids if a tuple is given.

This is to be in line with `cupy.cuda.nccl.NcclCommunicator.initAll()`: https://docs-cupy.chainer.org/en/latest/reference/generated/cupy.cuda.nccl.NcclCommunicator.html#cupy.cuda.nccl.NcclCommunicator.initAll

Reply:
The appropriate value of `use_multi_gpus` is uniquely determined from `devices`?

Reply:
I haven't made it so, but we can if you wish. In that case, users just specify `devices`, and with any meaningful `devices`,  the internal `use_multi_gpus` will be set to `True`. What do you think?

Reply:
@asi1024 How about this?

Reply:
Yes, please!